src/main/java/org/opensearch/knn/index/memory/NativeMemoryAllocation.java

```diff
@@ -0,0 +1,333 @@
+        private final long size;
+
+        // Implement reader/writer with semaphores to deal with passing lock conditions between threads
+        private int readCount;
+        private Semaphore readSemaphore;
+        private Semaphore writeSemaphore;
+
+        /**
+         * Constructor
+         *
+         * @param pointer pointer in memory to the training data allocation
+         * @param size amount memory needed for allocation in kilobytes
+         */
+        TrainingDataAllocation(long pointer, long size) {
+            this.closed = false;
+            this.pointer = pointer;
+            this.size = size;
+
+            this.readCount = 0;
+            this.readSemaphore = new Semaphore(1);
+            this.writeSemaphore = new Semaphore(1);
+        }
+
+        @Override
+        public void close() {
+            if (closed) {
+                return;
+            }
+
+            closed = true;
+            if (this.pointer != 0) {
+                JNIService.freeVectors(this.pointer);
+            }
+        }
+
+        @Override
+        public boolean isClosed() {
+            return closed;
+        }
+
+        @Override
+        public long getPointer() {
+            return pointer;
+        }
+
+        /**
+         * A read lock will be obtained when a training job needs access to the TrainingDataAllocation.
+         * In the future, we may want to switch to tryAcquire functionality.
+         */
+        @Override
+        public void readLock() {
+            try {
+                readSemaphore.acquire();
+            } catch (InterruptedException ex) {
+                throw new RuntimeException(ex);
+            }
+
+
+            if (readCount == 0) {
+                try {
+                    writeSemaphore.acquire();
+                } catch (InterruptedException e) {
+                    readSemaphore.release();
+                    throw new RuntimeException(e);
+                }
+            }
+
+            readCount++;
+            readSemaphore.release();
+        }
+
+        /**
+         * A write lock will be obtained either on eviction from {@link NativeMemoryCacheManager NativeMemoryManager's}
+         * or when training data is actually being loaded. A semaphore is used because collecting training data
+         * happens asynchrously, so the thread that obtains the lock will not be the same thread that releases the
+         * lock.
+         */
+        @Override
+        public void writeLock() {
+            try {
+                writeSemaphore.acquire();
+            } catch (InterruptedException ex) {
+                throw new RuntimeException(ex);
+            }
+        }
+
+        @Override
+        public void readUnlock() {
+            try {
+                readSemaphore.acquire();
+            } catch (InterruptedException ex) {
+                throw new RuntimeException(ex);
+            }
+
+            readCount--;
+
+            if (readCount == 0) {
```

Comment:
Minor: ```<=0``` just to be safe incase of multiple threads calling``` readunlock``` simultaneously?

Reply:
Sure, I can add that just to be safe.

This should never happen though. Because in line 300 the readSemaphore is obtained, other threads will block until it is released.